README.md

[![Build status](https://ci.appveyor.com/api/projects/status/fju06tfixss6dce8?svg=true)](https://ci.appveyor.com/project/masalmon/eml-tools) [![Build Status](https://travis-ci.org/maelle/eml.tools.svg?branch=master)](https://travis-ci.org/maelle/eml.tools) [![codecov](https://codecov.io/gh/maelle/eml.tools/branch/master/graph/badge.svg)](https://codecov.io/gh/maelle/eml.tools)

Package in construction for using EML (Ecological Metadata Language) in data analysis, currently only one function. For creating a new EML, see the [rOpenSci `EML` package](https://github.com/ropensci/EML). This package will take advantage from the fact that EML is machine readable to make some analysis steps easier. Any suggestion or contribution is welcome, but please note that this package has a [code of conduct](CONDUCT.md).

Translating factors into their meanings
=======================================

When the EML only has one table
-------------------------------

``` r
library("eml.tools")
data("animals_eml")
data("animals_data")
animals_data
```

    ##    animal age      size
    ## 1       A   A 50.902957
    ## 2       A   A 48.834863
    ## 3       A   A 51.521256
    ## 4       A   A 50.919211
    ## 5       A   A 50.801015
    ## 6       A   B 80.519204
    ## 7       A   B 80.889112
    ## 8       A   B 79.751250
    ## 9       A   B 80.310155
    ## 10      A   B 79.971646
    ## 11      B   A 10.609813
    ## 12      B   A  9.782217
    ## 13      B   A  9.753498
    ## 14      B   A 10.371456
    ## 15      B   A 10.487272
    ## 16      B   B 20.630848
    ## 17      B   B 19.361207
    ## 18      B   B 18.735406
    ## 19      B   B 19.214350
    ## 20      B   B 21.500681

``` r
eml_translate_factors(df = animals_data, eml = animals_eml, keep = FALSE)
```

    ## # A tibble: 20 × 3
    ##         size     animal      age
    ##        <dbl>      <chr>    <chr>
    ## 1  50.902957 monstercat juvenile
    ## 2  48.834863 monstercat juvenile
    ## 3  51.521256 monstercat juvenile
    ## 4  50.919211 monstercat juvenile
    ## 5  50.801015 monstercat juvenile
    ## 6  80.519204 monstercat    adult
    ## 7  80.889112 monstercat    adult
    ## 8  79.751250 monstercat    adult
    ## 9  80.310155 monstercat    adult
    ## 10 79.971646 monstercat    adult
    ## 11 10.609813 monsterdog juvenile
    ## 12  9.782217 monsterdog juvenile
    ## 13  9.753498 monsterdog juvenile
    ## 14 10.371456 monsterdog juvenile
    ## 15 10.487272 monsterdog juvenile
    ## 16 20.630848 monsterdog    adult
    ## 17 19.361207 monsterdog    adult
    ## 18 18.735406 monsterdog    adult
    ## 19 19.214350 monsterdog    adult
    ## 20 21.500681 monsterdog    adult

``` r
eml_translate_factors(df = animals_data, eml = animals_eml, keep = TRUE)
```

    ## # A tibble: 20 × 5
    ##    animal   age      size animal_meaning age_meaning
    ##     <chr> <chr>     <dbl>          <chr>       <chr>
    ## 1       A     A 50.902957     monstercat    juvenile
    ## 2       A     A 48.834863     monstercat    juvenile
    ## 3       A     A 51.521256     monstercat    juvenile
    ## 4       A     A 50.919211     monstercat    juvenile
    ## 5       A     A 50.801015     monstercat    juvenile
    ## 6       A     B 80.519204     monstercat       adult
    ## 7       A     B 80.889112     monstercat       adult
    ## 8       A     B 79.751250     monstercat       adult
    ## 9       A     B 80.310155     monstercat       adult
    ## 10      A     B 79.971646     monstercat       adult
    ## 11      B     A 10.609813     monsterdog    juvenile
    ## 12      B     A  9.782217     monsterdog    juvenile
    ## 13      B     A  9.753498     monsterdog    juvenile
    ## 14      B     A 10.371456     monsterdog    juvenile
    ## 15      B     A 10.487272     monsterdog    juvenile
    ## 16      B     B 20.630848     monsterdog       adult
    ## 17      B     B 19.361207     monsterdog       adult
    ## 18      B     B 18.735406     monsterdog       adult
    ## 19      B     B 19.214350     monsterdog       adult
    ## 20      B     B 21.500681     monsterdog       adult

When the EML only has several tables
------------------------------------

This is the case if the EML documents a database. In this case, we need to use the entityName of the table for which we want to perform the translation.

``` r
library("eml.tools")
data("animals_db_eml")
data("animals_data")
animals_data
```

    ## # A tibble: 20 × 3
    ##    animal    age      size
    ##    <fctr> <fctr>     <dbl>
    ## 1       A      A 50.902957
    ## 2       A      A 48.834863
    ## 3       A      A 51.521256
    ## 4       A      A 50.919211
    ## 5       A      A 50.801015
    ## 6       A      B 80.519204
    ## 7       A      B 80.889112
    ## 8       A      B 79.751250
    ## 9       A      B 80.310155
    ## 10      A      B 79.971646
    ## 11      B      A 10.609813
    ## 12      B      A  9.782217
    ## 13      B      A  9.753498
    ## 14      B      A 10.371456
    ## 15      B      A 10.487272
    ## 16      B      B 20.630848
    ## 17      B      B 19.361207
    ## 18      B      B 18.735406
    ## 19      B      B 19.214350
    ## 20      B      B 21.500681

``` r
eml_translate_factors(df = animals_data,
                      eml = animals_db_eml,
                      which_table = "english_animals",
                      keep = FALSE)
```

    ## # A tibble: 20 × 3
    ##         size     animal      age
    ##        <dbl>      <chr>    <chr>
    ## 1  50.902957 monstercat juvenile
    ## 2  48.834863 monstercat juvenile
    ## 3  51.521256 monstercat juvenile
    ## 4  50.919211 monstercat juvenile
    ## 5  50.801015 monstercat juvenile
    ## 6  80.519204 monstercat    adult
    ## 7  80.889112 monstercat    adult
    ## 8  79.751250 monstercat    adult
    ## 9  80.310155 monstercat    adult
    ## 10 79.971646 monstercat    adult
    ## 11 10.609813 monsterdog juvenile
    ## 12  9.782217 monsterdog juvenile
    ## 13  9.753498 monsterdog juvenile
    ## 14 10.371456 monsterdog juvenile
    ## 15 10.487272 monsterdog juvenile
    ## 16 20.630848 monsterdog    adult
    ## 17 19.361207 monsterdog    adult
    ## 18 18.735406 monsterdog    adult
    ## 19 19.214350 monsterdog    adult
    ## 20 21.500681 monsterdog    adult

``` r
eml_translate_factors(df = animals_data,
                      eml = animals_db_eml,
                      which_table = "english_animals",
                      keep = TRUE)
```

    ## # A tibble: 20 × 5
    ##    animal   age      size animal_meaning age_meaning
    ##     <chr> <chr>     <dbl>          <chr>       <chr>
    ## 1       A     A 50.902957     monstercat    juvenile
    ## 2       A     A 48.834863     monstercat    juvenile
    ## 3       A     A 51.521256     monstercat    juvenile
    ## 4       A     A 50.919211     monstercat    juvenile
    ## 5       A     A 50.801015     monstercat    juvenile
    ## 6       A     B 80.519204     monstercat       adult
    ## 7       A     B 80.889112     monstercat       adult
    ## 8       A     B 79.751250     monstercat       adult
    ## 9       A     B 80.310155     monstercat       adult
    ## 10      A     B 79.971646     monstercat       adult
    ## 11      B     A 10.609813     monsterdog    juvenile
    ## 12      B     A  9.782217     monsterdog    juvenile
    ## 13      B     A  9.753498     monsterdog    juvenile
    ## 14      B     A 10.371456     monsterdog    juvenile
    ## 15      B     A 10.487272     monsterdog    juvenile
    ## 16      B     B 20.630848     monsterdog       adult
    ## 17      B     B 19.361207     monsterdog       adult
    ## 18      B     B 18.735406     monsterdog       adult
    ## 19      B     B 19.214350     monsterdog       adult
    ## 20      B     B 21.500681     monsterdog       adult


NEWS.md
# eml.tools 0.1.0

* Added the eml_translate_factors function.

* Added a `NEWS.md` file to track changes to the package.





tests/testthat.R
library(testthat)
library(eml.tools)

test_check("eml.tools")


tests/testthat/test-eml_translate_factors.R
context("eml_translate_factors")

test_that("eml_translate_factors works for one table", {
  data("animals_eml")
  data("animals_data")
  output1 <- eml_translate_factors(df = animals_data, eml = animals_eml, keep = FALSE)
  expect_is(output1, "tbl_df")
  expect_equal(ncol(output1), ncol(animals_data))
  expect_equal(sum(is.na(output1$age)), 0)
  expect_equal(sum(is.na(output1$animal)), 0)
  output2 <- eml_translate_factors(df = animals_data,
                                   eml = animals_eml,
                                   keep = TRUE)
  expect_is(output2, "tbl_df")
  expect_equal(ncol(output2), ncol(animals_data) + 2)
  expect_equal(sum(is.na(output2$age_meaning)), 0)
  expect_equal(sum(is.na(output2$animal_meaning)), 0)
})



test_that("eml_translate_factors works for a group of tables", {
  data("animals_db_eml")
  data("animals_data")
  output1 <- eml_translate_factors(df = animals_data,
                                   eml = animals_db_eml,
                                   keep = FALSE,
                                   which_table = "english_animals")
  expect_is(output1, "tbl_df")
  expect_equal(ncol(output1), ncol(animals_data))
  expect_equal(sum(is.na(output1$age)), 0)
  expect_equal(sum(is.na(output1$animal)), 0)
  output2 <- eml_translate_factors(df = animals_data,
                                   eml = animals_db_eml,
                                   keep = TRUE,
                                   which_table = "english_animals")
  expect_is(output2, "tbl_df")
  expect_equal(ncol(output2), ncol(animals_data) + 2)
  expect_equal(sum(is.na(output2$age_meaning)), 0)
  expect_equal(sum(is.na(output2$animal_meaning)), 0)
})


R/data_doc.R
#' Example data.frame with fake animals data
#'
#' Example data.frame with fake animals data.
#'
#' A data.frame(\code{dplyr tbl_df}) with
#' \itemize{
#' \item animal. \code{factor}, animal code
#' \item age \code{factor}, age category code
#' \item size \code{numeric}, size of the animal
#' }
#'
#' @format A data.frame(\code{dplyr tbl_df})
#' @name animals_data
#' @docType data
#' @keywords data
#' @usage data(animals_data)
NULL

#' Example eml for the fake animals data
#'
#' Example eml for the fake animals data
#'
#' An \code{eml}
#'
#' @format \code{eml}
#' @name animals_eml
#' @docType data
#' @keywords data
#' @usage data(animals_eml)
NULL

#' Other example eml for the fake animals data
#'
#' Other example eml for the fake animals data
#'
#' An \code{eml} for a dataset, not a single table.
#'
#' @format \code{eml}
#' @name animals_db_eml
#' @docType data
#' @keywords data
#' @usage data(animals_db_eml)
NULL


inst/example_data.R
library("EML")

# create the example data
animals_data <- data.frame(animal = c(rep("A", 10), rep("B", 10)),
                   age = c(rep("A", 5), rep("B", 5), rep("A", 5), rep("B", 5)),
                   size = c(rnorm(mean = 50, n = 5),
                            rnorm(mean = 80, n = 5),
                            rnorm(mean = 10, n = 5),
                            rnorm(mean = 20, n = 5)))
animals_data <- tibble::as_tibble(animals_data)
save(animals_data, file = "data/animals_data.RData")

# attributesList
attributes <-
  data.frame(
    attributeName = c(
      "animal",
      "age",
      "size"),
    attributeDefinition = c(
      "animal species",
      "life stage",
      "body length"),
    formatString = c(
      NA,
      NA,
      NA),
    definition = c(
      "animal species",
      "life stage",
      "body length"),
    unit = c(
      NA,
      NA,
      "meter"),
    numberType = c(
      NA,
      NA,
      "real"),
    stringsAsFactors = FALSE
  )

# two of the attributes are factors
animal.codes <- c(A = "monstercat",
                  B = "monsterdog")
age.codes <- c(A = "juvenile",
               B = "adult")
factors <- rbind(
  data.frame(
    attributeName = "animal",
    code = names(animal.codes),
    definition = unname(animal.codes)
  ),
  data.frame(
    attributeName = "age",
    code = names(age.codes),
    definition = unname(age.codes)
  ))


attributeList <- set_attributes(attributes,
                                factors,
                                col_classes = c("factor", "factor", "numeric"))

# physical
physical <- set_physical("doesntexist.csv")

# datatable
dataTable <- new("dataTable",
                 entityName = "doesntexist.csv",
                 entityDescription = "tipping point experiment 1",
                 physical = physical,
                 attributeList = attributeList)

# coverage

geographicDescription <- "Nowhere real, thankfully given the size of these animals."


coverage <-
  set_coverage(begin = '2016-10-01', end = '2016-10-31',
               sci_names = "Monstri monstri",
               geographicDescription = geographicDescription,
               west = 0, east = 0,
               north = 0, south = 0,
               altitudeMin = 160, altitudeMaximum = 330,
               altitudeUnits = "meter")

# abstract & co
pubDate <- "2016"

title <- "Test dataset"

abstract <- "Test dataset and eml to write a function for translating factors"

intellectualRights <- "Free to use"

# parties
R_person <- as.person("The Dataset Creator [cre] <fakeaddress@email.com>")
creator <- as(R_person, "creator")

associatedParty <- as.person("Another Contributor [ctb] <otherfakeadress@email.com>")
associatedParty <- as(associatedParty, "associatedParty")

# methods
methods_file <- paste0(getwd(), "/inst/extdata/methods.txt")
methods <- set_methods(methods_file)

# contact
HF_address <- new("address",
                  deliveryPoint = "324 North Main Street",
                  city = "Petersham",
                  administrativeArea = "MA",
                  postalCode = "01366",
                  country = "USA")
publisher <- new("publisher",
                 organizationName = "Harvard Forest",
                 address = HF_address)
contact <-
  new("contact",
      individualName = creator@individualName,
      electronicMail = creator@electronicMailAddress,
      address = HF_address,
      organizationName = "Harvard Forest",
      phone = "000-000-0000")

# dataset

dataset <- new("dataset",
               title = title,
               creator = creator,
               pubDate = pubDate,
               intellectualRights = intellectualRights,
               abstract = abstract,
               associatedParty = associatedParty,
               coverage = coverage,
               contact = contact,
               methods = methods,
               dataTable = dataTable)

# the function should return column names that look like the code column names,
# not definition for all!


animals_eml <- new("eml",
           packageId = "is",
           system = "uuid", # type of identifier
           dataset = dataset)
eml_validate(animals_eml)
write_eml(animals_eml, file = paste0(getwd(), "/inst/extdata/example.xml"))
save(animals_eml, file = "data/animals_eml.RData")


R/eml_translate_factors.R
#' Translate factors into their meanings
#'
#' @param df Data frame
#' @param eml eml metadata
#' @param keep Whether to keep the columns with codes. If FALSE, the meanings replace the factors previous codes.
#' @param which_table for which table to retrieve the factors, if there are several tables (entityName)
#'
#' @return A data.frame with translated factors
#' @export
#'
#' @examples
#' data("animals_eml")
#' data("animals_data")
#' eml_translate_factors(df = animals_data, eml = animals_eml, keep = FALSE)
#' eml_translate_factors(df = animals_data, eml = animals_eml, keep = TRUE)
#'
eml_translate_factors <- function(df, eml, keep = FALSE,
                                  which_table = NULL){

  if(is.null(which_table)){
    df_factor <- as.data.frame(EML::eml_get(eml, "attributeList")$factors)

  }else{

    dt <- EML::eml_get(eml, "dataTable")

    names <- vapply(dt, get_name, "")
    which_table <- which(names == which_table)
    attrs <- EML::eml_get(dt[[which_table]], "attributeList")
    df_factor <- as.data.frame(attrs$factors)
  }
  names_factors <- unique(df_factor$attributeName)
  names_factors <- names_factors[names_factors %in% names(df)]
  names_factors <- as.character(names_factors)

  for(name in names_factors){
    df <- eml_translate_one_factor(name, df = df,
                                   df_factor = df_factor,
                                   keep = keep)
  }

  return(df)
}

eml_translate_one_factor <- function(name, df, df_factor, keep){
  df_factor <- dplyr::filter_(df_factor, lazyeval::interp(~ attributeName == name))
  df_factor <- dplyr::select_(df_factor, quote(- attributeName))

  if(keep){
    df[[name]] <- as.character(df[[name]])
    df_factor <- dplyr::rename_(df_factor, .dots=setNames("definition", paste0(name, "_meaning")))
    df <- suppressWarnings(dplyr::left_join(df, df_factor, by = setNames("code", name)))
  }else{
    df_factor <- dplyr::rename_(df_factor, .dots=setNames("definition", name))
    df <- dplyr::rename_(df, .dots=setNames(name, "code"))
    df <- dplyr::mutate_(df, code = lazyeval::interp(~ as.character(code)))
    df <- suppressWarnings(dplyr::left_join(df, df_factor, by = setNames("code", "code")))
    df <- dplyr::select_(df, quote(- code))
  }
  return(df)
}


get_name <- function(table){
  EML::eml_get(table, "entityName")
}


inst/example_data_several_tables.R
library("EML")

# attributesList
attributes <-
  data.frame(
    attributeName = c(
      "animal",
      "age",
      "size"),
    attributeDefinition = c(
      "animal species",
      "life stage",
      "body length"),
    formatString = c(
      NA,
      NA,
      NA),
    definition = c(
      "animal species",
      "life stage",
      "body length"),
    unit = c(
      NA,
      NA,
      "meter"),
    numberType = c(
      NA,
      NA,
      "real"),
    stringsAsFactors = FALSE
  )

# two of the attributes are factors
animal.codes <- c(A = "monstercat",
                  B = "monsterdog")
age.codes <- c(A = "juvenile",
               B = "adult")
factors1 <- rbind(
  data.frame(
    attributeName = "animal",
    code = names(animal.codes),
    definition = unname(animal.codes)
  ),
  data.frame(
    attributeName = "age",
    code = names(age.codes),
    definition = unname(age.codes)
  ))


attributeList1 <- set_attributes(attributes,
                                 factors1,
                                 col_classes = c("factor", "factor", "numeric"))
animal.codes <- c(A = "monstercat",
                  B = "monsterdog")
age.codes <- c(A = "petit",
               B = "grand")
factors2 <- rbind(
  data.frame(
    attributeName = "animal",
    code = names(animal.codes),
    definition = unname(animal.codes)
  ),
  data.frame(
    attributeName = "age",
    code = names(age.codes),
    definition = unname(age.codes)
  ))


attributeList2 <- set_attributes(attributes,
                                 factors2,
                                 col_classes = c("factor", "factor", "numeric"))

# physical
physical <- set_physical("doesntexist.csv")

# datatable
dataTable1 <- new("dataTable",
                  entityName = "english_animals",
                  entityDescription = "tipping point experiment 1",
                  physical = physical,
                  attributeList = attributeList1)
dataTable2 <- new("dataTable",
                  entityName = "french_animals",
                  entityDescription = "tipping point experiment 1",
                  physical = physical,
                  attributeList = attributeList2)

# coverage

geographicDescription <- "Nowhere real, thankfully given the size of these animals."


coverage <-
  set_coverage(begin = '2016-10-01', end = '2016-10-31',
               sci_names = "Monstri monstri",
               geographicDescription = geographicDescription,
               west = 0, east = 0,
               north = 0, south = 0,
               altitudeMin = 160, altitudeMaximum = 330,
               altitudeUnits = "meter")

# abstract & co
pubDate <- "2016"

title <- "Test dataset"

abstract <- "Test dataset and eml to write a function for translating factors"

intellectualRights <- "Free to use"

# parties
R_person <- as.person("The Dataset Creator [cre] <fakeaddress@email.com>")
creator <- as(R_person, "creator")

associatedParty <- as.person("Another Contributor [ctb] <otherfakeadress@email.com>")
associatedParty <- as(associatedParty, "associatedParty")

# methods
methods_file <- paste0(getwd(), "/inst/extdata/methods.txt")
methods <- set_methods(methods_file)

# contact
HF_address <- new("address",
                  deliveryPoint = "324 North Main Street",
                  city = "Petersham",
                  administrativeArea = "MA",
                  postalCode = "01366",
                  country = "USA")
publisher <- new("publisher",
                 organizationName = "Harvard Forest",
                 address = HF_address)
contact <-
  new("contact",
      individualName = creator@individualName,
      electronicMail = creator@electronicMailAddress,
      address = HF_address,
      organizationName = "Harvard Forest",
      phone = "000-000-0000")

# dataset

dataset <- new("dataset",
               title = title,
               creator = creator,
               pubDate = pubDate,
               intellectualRights = intellectualRights,
               abstract = abstract,
               associatedParty = associatedParty,
               coverage = coverage,
               contact = contact,
               methods = methods,
               dataTable = list(dataTable1,
                                dataTable2))

# the function should return column names that look like the code column names,
# not definition for all!


animals_db_eml <- new("eml",
                   packageId = "is",
                   system = "uuid", # type of identifier
                   dataset = dataset)
eml_validate(animals_db_eml)
write_eml(animals_db_eml, file = paste0(getwd(), "/inst/extdata/db_example.xml"))
save(animals_db_eml, file = "data/animals_db_eml.RData")
